File: sdks/go/filter_test.go
package vectorizer

import (
	"encoding/json"
	"testing"
)

// marshalJSON is a test helper that marshals v to a generic map for assertion.
func marshalJSON(t *testing.T, v interface{}) map[string]interface{} {
	t.Helper()
	b, err := json.Marshal(v)
	if err != nil {
		t.Fatalf("json.Marshal failed: %v", err)
	}
	var m map[string]interface{}
	if err := json.Unmarshal(b, &m); err != nil {
		t.Fatalf("json.Unmarshal failed: %v", err)
	}
	return m
}

// TestFilterEqWireShape verifies that FilterEq serialises to the expected
// wire shape: {"key":"topic","match":{"value":"index"}}.
func TestFilterEqWireShape(t *testing.T) {
	cond := FilterEq("topic", "index")
	m := marshalJSON(t, cond)

	if m["key"] != "topic" {
		t.Errorf("key should be 'topic' but got %v", m["key"])
	}
	match, ok := m["match"].(map[string]interface{})
	if !ok {
		t.Fatalf("match should be an object but got %T", m["match"])
	}
	if match["value"] != "index" {
		t.Errorf("match.value should be 'index' but got %v", match["value"])
	}
	if _, exists := m["range"]; exists {
		t.Error("range key should be absent for FilterEq")
	}
	if _, exists := m["filter"]; exists {
		t.Error("filter key should be absent for FilterEq")
	}
}

// TestFilterInWireShape verifies that FilterIn serialises to
// {"key":"tier","match":{"any":["hot","warm"]}}.
func TestFilterInWireShape(t *testing.T) {
	cond := FilterIn("tier", []interface{}{"hot", "warm"})
	m := marshalJSON(t, cond)

	if m["key"] != "tier" {
		t.Errorf("key should be 'tier' but got %v", m["key"])
	}
	match, ok := m["match"].(map[string]interface{})
	if !ok {
		t.Fatalf("match should be an object but got %T", m["match"])
	}
	anyArr, ok := match["any"].([]interface{})
	if !ok {
		t.Fatalf("match.any should be an array but got %T", match["any"])
	}
	if len(anyArr) != 2 {
		t.Fatalf("match.any should have 2 elements but got %d", len(anyArr))
	}
	if anyArr[0] != "hot" {
		t.Errorf("match.any[0] should be 'hot' but got %v", anyArr[0])
	}
	if anyArr[1] != "warm" {
		t.Errorf("match.any[1] should be 'warm' but got %v", anyArr[1])
	}
	if _, exists := match["value"]; exists {
		t.Error("match.value key should be absent for FilterIn")
	}
}

// TestFilterRangeWireShape verifies that FilterRangeGteLte serialises to
// {"key":"score","range":{"gte":0.5,"lte":0.9}}.
func TestFilterRangeWireShape(t *testing.T) {
	gte := 0.5
	lte := 0.9
	cond := FilterRangeGteLte("score", &gte, &lte)
	m := marshalJSON(t, cond)

	if m["key"] != "score" {
		t.Errorf("key should be 'score' but got %v", m["key"])
	}
	rng, ok := m["range"].(map[string]interface{})
	if !ok {
		t.Fatalf("range should be an object but got %T", m["range"])
	}
	if rng["gte"] != 0.5 {
		t.Errorf("range.gte should be 0.5 but got %v", rng["gte"])
	}
	if rng["lte"] != 0.9 {
		t.Errorf("range.lte should be 0.9 but got %v", rng["lte"])
	}
	if _, exists := m["match"]; exists {
		t.Error("match key should be absent for FilterRangeGteLte")
	}
}

// TestQdrantFilterMustOnly verifies that MustFilter serialises with a "must"
// key and omits the empty "should" and "must_not" keys.
func TestQdrantFilterMustOnly(t *testing.T) {
	f := MustFilter(FilterEq("topic", "index"))
	b, err := json.Marshal(f)
	if err != nil {
		t.Fatalf("json.Marshal failed: %v", err)
	}
	var m map[string]interface{}
	if err := json.Unmarshal(b, &m); err != nil {
		t.Fatalf("json.Unmarshal failed: %v", err)
	}

	if _, ok := m["must"]; !ok {
		t.Error("must key should be present")
	}
	if _, ok := m["should"]; ok {
		t.Error("should key should be absent when empty")
	}
	if _, ok := m["must_not"]; ok {
		t.Error("must_not key should be absent when empty")
	}

	mustArr, ok := m["must"].([]interface{})
	if !ok || len(mustArr) == 0 {
		t.Fatalf("must should be a non-empty array but got %v", m["must"])
	}
	cond, ok := mustArr[0].(map[string]interface{})
	if !ok {
		t.Fatalf("must[0] should be an object but got %T", mustArr[0])
	}
	if cond["key"] != "topic" {
		t.Errorf("must[0].key should be 'topic' but got %v", cond["key"])
	}
}

// TestQdrantFilterCompound verifies that a filter with both must and must_not
// conditions serialises both keys while omitting the empty should key.
func TestQdrantFilterCompound(t *testing.T) {
	f := QdrantFilter{
		Must:    []FilterCondition{FilterEq("status", "active")},
		MustNot: []FilterCondition{FilterEq("tier", "cold")},
	}
	b, err := json.Marshal(f)
	if err != nil {
		t.Fatalf("json.Marshal failed: %v", err)
	}
	var m map[string]interface{}
	if err := json.Unmarshal(b, &m); err != nil {
		t.Fatalf("json.Unmarshal failed: %v", err)
	}

	if _, ok := m["must"]; !ok {
		t.Error("must key should be present")
	}
	if _, ok := m["must_not"]; !ok {
		t.Error("must_not key should be present")
	}
	if _, ok := m["should"]; ok {
		t.Error("should key should be absent when empty")
	}
}

// TestNestedFilter verifies that FilterNested produces a condition with a
// "filter" key containing the sub-filter object.
func TestNestedFilter(t *testing.T) {
	inner := MustFilter(FilterEq("inner_key", "value"))
	cond := FilterNested(inner)
	m := marshalJSON(t, cond)

	filterObj, ok := m["filter"].(map[string]interface{})
	if !ok {
		t.Fatalf("filter key should be an object but got %T", m["filter"])
	}
	mustArr, ok := filterObj["must"].([]interface{})
	if !ok || len(mustArr) == 0 {
		t.Fatalf("nested filter.must should be non-empty but got %v", filterObj["must"])
	}
	innerCond, ok := mustArr[0].(map[string]interface{})
	if !ok {
		t.Fatalf("nested must[0] should be an object but got %T", mustArr[0])
	}
	if innerCond["key"] != "inner_key" {
		t.Errorf("nested must[0].key should be 'inner_key' but got %v", innerCond["key"])
	}
}

// TestIsEmpty verifies the IsEmpty method covers all cases.
func TestIsEmpty(t *testing.T) {
	// Zero-value struct must be empty.
	var zero QdrantFilter
	if !zero.IsEmpty() {
		t.Error("zero-value QdrantFilter should be empty")
	}

	// Struct with all explicit empty slices must be empty.
	withEmptySlices := QdrantFilter{
		Must:    []FilterCondition{},
		Should:  []FilterCondition{},
		MustNot: []FilterCondition{},
	}
	if !withEmptySlices.IsEmpty() {
		t.Error("QdrantFilter with all empty slices should be empty")
	}

	// Struct with one populated Must array must NOT be empty.
	withMust := MustFilter(FilterEq("k", "v"))
	if withMust.IsEmpty() {
		t.Error("QdrantFilter with must conditions should not be empty")
	}

	// Struct with one populated Should array must NOT be empty.
	withShould := ShouldFilter(FilterEq("k", "v"))
	if withShould.IsEmpty() {
		t.Error("QdrantFilter with should conditions should not be empty")
	}

	// Struct with one populated MustNot array must NOT be empty.
	withMustNot := MustNotFilter(FilterEq("k", "v"))
	if withMustNot.IsEmpty() {
		t.Error("QdrantFilter with must_not conditions should not be empty")
	}
}

File: sdks/typescript/tests/filter.test.ts
/**
 * Unit tests for the typed QdrantFilter builder (phase23 §4).
 *
 * Verifies:
 * - Wire JSON produced by each builder helper matches the server's expected shape.
 * - Compound must + must_not produces the correct multi-clause object.
 * - Empty-filter guard (isNonEmptyFilter) rejects filters with no conditions.
 * - Builder ergonomics: filter.combine() assembles multi-clause filters.
 *
 * Wire shapes reference: sdks/rust/src/models/filter.rs
 */

import { describe, it, expect } from 'vitest';
import {
  filter,
  isNonEmptyFilter,
  type QdrantFilter,
  type FilterCondition,
} from '../src/models/filter';

// ---------------------------------------------------------------------------
// Single-condition builders → correct wire JSON
// ---------------------------------------------------------------------------

describe('filter.eq — exact match wire shape', () => {
  it('produces {key, match: {value}} for a string value', () => {
    const cond: FilterCondition = filter.eq('topic', 'index');
    expect(cond.key).toBe('topic');
    expect(cond.match).toBeDefined();
    expect(cond.match!.value).toBe('index');
    expect(cond.match!.any).toBeUndefined();
    expect(cond.range).toBeUndefined();
    expect(cond.filter).toBeUndefined();

    // Serialise to JSON and verify the wire shape
    const json = JSON.parse(JSON.stringify(cond));
    expect(json).toEqual({ key: 'topic', match: { value: 'index' } });
  });

  it('produces the correct wire shape for a numeric value', () => {
    const cond = filter.eq('count', 42);
    const json = JSON.parse(JSON.stringify(cond));
    expect(json).toEqual({ key: 'count', match: { value: 42 } });
  });

  it('produces the correct wire shape for a boolean value', () => {
    const cond = filter.eq('active', true);
    const json = JSON.parse(JSON.stringify(cond));
    expect(json).toEqual({ key: 'active', match: { value: true } });
  });
});

describe('filter.in — membership match wire shape', () => {
  it('produces {key, match: {any: [...]}} for an array', () => {
    const cond = filter.in('status', ['hot', 'warm', 'cold']);
    const json = JSON.parse(JSON.stringify(cond));
    expect(json).toEqual({ key: 'status', match: { any: ['hot', 'warm', 'cold'] } });
  });

  it('handles a single-element array', () => {
    const cond = filter.in('tier', ['hot']);
    const json = JSON.parse(JSON.stringify(cond));
    expect(json).toEqual({ key: 'tier', match: { any: ['hot'] } });
  });
});

describe('filter.range — numeric range wire shape', () => {
  it('produces {key, range: {gte, lte}} when both bounds are set', () => {
    const cond = filter.range('score', { gte: 0.5, lte: 0.9 });
    const json = JSON.parse(JSON.stringify(cond));
    expect(json).toEqual({ key: 'score', range: { gte: 0.5, lte: 0.9 } });
  });

  it('omits absent bound keys (gte-only)', () => {
    const cond = filter.range('score', { gte: 0.8 });
    const json = JSON.parse(JSON.stringify(cond));
    expect(json).toEqual({ key: 'score', range: { gte: 0.8 } });
    expect(json.range.lte).toBeUndefined();
  });

  it('omits absent bound keys (lte-only)', () => {
    const cond = filter.range('age', { lte: 100 });
    const json = JSON.parse(JSON.stringify(cond));
    expect(json).toEqual({ key: 'age', range: { lte: 100 } });
    expect(json.range.gte).toBeUndefined();
  });
});

// ---------------------------------------------------------------------------
// Compound filter builders
// ---------------------------------------------------------------------------

describe('filter.must — AND compound filter', () => {
  it('wraps conditions in a must array', () => {
    const f: QdrantFilter = filter.must(filter.eq('tier', 'hot'));
    const json = JSON.parse(JSON.stringify(f));
    expect(json.must).toHaveLength(1);
    expect(json.must[0]).toEqual({ key: 'tier', match: { value: 'hot' } });
    expect(json.should).toBeUndefined();
    expect(json.must_not).toBeUndefined();
  });

  it('accepts multiple conditions as spread args', () => {
    const f = filter.must(filter.eq('tier', 'hot'), filter.range('score', { gte: 0.8 }));
    const json = JSON.parse(JSON.stringify(f));
    expect(json.must).toHaveLength(2);
  });
});

describe('filter.mustNot — NOT compound filter', () => {
  it('wraps conditions in a must_not array', () => {
    const f = filter.mustNot(filter.eq('archived', true));
    const json = JSON.parse(JSON.stringify(f));
    expect(json.must_not).toHaveLength(1);
    expect(json.must).toBeUndefined();
    expect(json.should).toBeUndefined();
  });
});

describe('compound must + must_not — correct wire JSON', () => {
  it('produces both clauses in one filter object', () => {
    const f: QdrantFilter = filter.combine({
      must: [filter.eq('tier', 'hot'), filter.range('score', { gte: 0.8 })],
      must_not: [filter.eq('archived', true)],
    });

    const json = JSON.parse(JSON.stringify(f));
    expect(json.must).toHaveLength(2);
    expect(json.must[0]).toEqual({ key: 'tier', match: { value: 'hot' } });
    expect(json.must[1]).toEqual({ key: 'score', range: { gte: 0.8 } });
    expect(json.must_not).toHaveLength(1);
    expect(json.must_not[0]).toEqual({ key: 'archived', match: { value: true } });
    expect(json.should).toBeUndefined();
  });
});

// ---------------------------------------------------------------------------
// Empty-filter guard
// ---------------------------------------------------------------------------

describe('isNonEmptyFilter — rejects empty filters', () => {
  it('returns false for an object with no clauses', () => {
    expect(isNonEmptyFilter({})).toBe(false);
  });

  it('returns false when all clause arrays are empty', () => {
    expect(isNonEmptyFilter({ must: [], should: [], must_not: [] })).toBe(false);
  });

  it('returns false when arrays are null-ish via cast', () => {
    // Undefined arrays are treated as absent
    expect(isNonEmptyFilter({ must: undefined, should: undefined })).toBe(false);
  });

  it('returns true when must has at least one condition', () => {
    expect(isNonEmptyFilter({ must: [filter.eq('k', 'v')] })).toBe(true);
  });

  it('returns true when should has at least one condition', () => {
    expect(isNonEmptyFilter({ should: [filter.eq('k', 'v')] })).toBe(true);
  });

  it('returns true when must_not has at least one condition', () => {
    expect(isNonEmptyFilter({ must_not: [filter.eq('k', 'v')] })).toBe(true);
  });
});

// ---------------------------------------------------------------------------
// Builder ergonomics — chained combine
// ---------------------------------------------------------------------------

describe('filter.combine — multi-clause assembly', () => {
  it('assembles must + should + must_not in one call', () => {
    const f = filter.combine({
      must: [filter.eq('type', 'article')],
      should: [filter.in('tag', ['ai', 'ml'])],
      must_not: [filter.eq('draft', true)],
    });

    const json = JSON.parse(JSON.stringify(f));
    expect(json.must).toHaveLength(1);
    expect(json.should).toHaveLength(1);
    expect(json.must_not).toHaveLength(1);
    expect(json.should[0]).toEqual({ key: 'tag', match: { any: ['ai', 'ml'] } });
  });
});

// ---------------------------------------------------------------------------
// Nested sub-filter
// ---------------------------------------------------------------------------

describe('filter.nested — nested sub-filter condition', () => {
  it('produces a condition with a filter sub-object', () => {
    const inner = filter.must(filter.eq('inner_key', 'value'));
    const cond = filter.nested(inner);
    const json = JSON.parse(JSON.stringify(cond));
    expect(json.key).toBe('__nested__');
    expect(json.filter).toBeDefined();
    expect(json.filter.must[0]).toEqual({ key: 'inner_key', match: { value: 'value' } });
  });
});
